lib/storage.js

```diff
@@ -13,47 +13,98 @@
+const SOURCE_BASENAME = 'source';
 
-async function getSource(paramsSource, inDirectory, disableSourceDownload) {
-    // Note: validation has happened in validate.js before
+const DATA_PROTOCOL = 'data:';
 
+async function getAsset(assetReference, directory, name, disableDownload) {
+    // normalize asset reference to be an object
+    if(!assetReference) {
+        throw new GenericError('Missing assetReference');
+    }
+    if (typeof assetReference === 'string') {
+        assetReference = { url: assetReference };
+    }
     if (process.env.WORKER_TEST_MODE) {
         // local file support for `Asset Compute test-worker` unit tests
         // not supported for clients in production
         // params.source.url will just be a filename like 'file.jpg'
-        const filePath = path.join(inDirectory, paramsSource.url);
+        name = assetReference.url;
+        const filePath = path.join(directory, name);
         if (!fs.existsSync(filePath)) {
-            throw new Error(`Invalid or missing local file ${paramsSource.url}`);
+            throw new Error(`Invalid or missing local file ${name}`);
         }
-        return new Source(paramsSource.url, inDirectory);
+        return new Source(assetReference, directory, name);
 
     } else {
-        const source = new Source(paramsSource, inDirectory);
-        if (disableSourceDownload) {
+        const source = new Source(assetReference, directory, name);
+        if (disableDownload) {
             console.log(`Skipping source file download for ${AssetComputeLogUtils.redactUrl(source.url)}`);
         } else {
             // TODO: error handling: errors catched & thrown in http.download() and then catched again in worker.js ???
             // download http/https url into file
-            const protocol = URL.parse(paramsSource.url).protocol;
+            const protocol = URL.parse(assetReference.url).protocol;
             if (protocol === DATA_PROTOCOL) {
-                await datauri.download(paramsSource, source.path);
+                await datauri.download(assetReference, source.path);
             } else {
-                await http.download(paramsSource, source.path);
+                await http.download(assetReference, source.path);
             }
         }
 
         return source;
     }
 }
 
+// function to return an extension for a file
+// if not empty returns a leading period
+// prefers extension from the file over name determined by mimeType
+function extension(filename, mimeType) {
+    let ext = '';
+    if (filename) {
+        ext = path.extname(filename);
+    }
+    if (!ext && mimeType) {
+        const mimeExt = mime.extension(mimeType);
+        ext = mimeExt ? `.${mimeExt}` : '';
+    }
+    return ext;
+}
+
+// There is at least one worker (graphics magick) that in some cases depends
+// upon the file extension so it is best to try to use the appropriate one
+// based on the filename, url, or mimetype
+function sourceFilename(source) {
```

Comment:
Given the comment, is this still needed?

Reply:
Not quite I think. We talked about it in the mob but we didn't want to create more risks by changing this logic. We can look at it after watermarking is done.

One little benefit is that we see the source type in logs if it says `source.png` instead of just `source`. But not sure if that's important, the "real" source info is also available in the logs.